File: web/src/lib/monitoring/sentry.ts
/**
 * Sentry Error Tracking Configuration
 *
 * Centralized error tracking and monitoring using Sentry.io
 * Captures errors, performance metrics, and user context.
 *
 * Setup:
 * 1. Sign up at https://sentry.io (Free tier: 5K events/month)
 * 2. Create project and get DSN
 * 3. Add NEXT_PUBLIC_SENTRY_DSN to .env.local
 * 4. Call initSentry() in your root layout
 */

import * as Sentry from '@sentry/nextjs'

/**
 * Initialize Sentry for error tracking
 * Call this in your root layout or _app file
 */
export function initSentry() {
  const dsn = process.env.NEXT_PUBLIC_SENTRY_DSN || process.env.SENTRY_DSN

  if (!dsn) {
    console.warn('[Sentry] DSN not configured, error tracking disabled')
    return
  }

  Sentry.init({
    dsn,
    environment: process.env.NODE_ENV || 'development',
    tracesSampleRate: process.env.NODE_ENV === 'production' ? 0.1 : 1.0,

    // Session replay for debugging user issues
    replaysSessionSampleRate: 0.1, // 10% of sessions
    replaysOnErrorSampleRate: 1.0, // 100% of sessions with errors

    // Performance monitoring
    profilesSampleRate: process.env.NODE_ENV === 'production' ? 0.1 : 1.0,

    // Filter out sensitive data
    beforeSend(event, hint) {
      // Remove sensitive data from event
      if (event.request) {
        // Remove authorization headers
        if (event.request.headers) {
          delete event.request.headers['authorization']
          delete event.request.headers['cookie']
        }

        // Remove sensitive query parameters
        if (event.request.query_string) {
          const sanitized = event.request.query_string.replace(
            /token=[^&]*/gi,
            'token=[REDACTED]'
          )
          event.request.query_string = sanitized
        }
      }

      // Remove sensitive context data
      if (event.contexts) {
        if (event.contexts.stripe) {
          delete event.contexts.stripe.secretKey
        }
      }

      return event
    },

    // Ignore common non-critical errors
    ignoreErrors: [
      // Browser extensions
      'top.GLOBALS',
      'originalCreateNotification',
      'canvas.contentDocument',
      'MyApp_RemoveAllHighlights',
      'atomicFindClose',

      // Network errors that are expected
      'NetworkError',
      'Network request failed',
      'Failed to fetch',

      // User cancelled actions
      'AbortError',
      'The user aborted a request',
    ],

    // Only track important breadcrumbs
    beforeBreadcrumb(breadcrumb, hint) {
      // Don't track console logs in production
      if (breadcrumb.category === 'console' && process.env.NODE_ENV === 'production') {
        return null
      }

      // Don't track XHR to tracking services
      if (breadcrumb.category === 'xhr' && breadcrumb.data?.url) {
        const url = breadcrumb.data.url
        if (url.includes('analytics') || url.includes('tracking')) {
          return null
        }
      }

      return breadcrumb
    },
  })
}

/**
 * Capture an error with additional context
 */
export function captureError(
  error: Error,
  context?: {
    user?: { id: string; email?: string; username?: string }
    tags?: Record<string, string>
    extra?: Record<string, any>
    level?: Sentry.SeverityLevel
  }
) {
  if (context) {
    Sentry.withScope((scope) => {
      if (context.user) {
        scope.setUser(context.user)
      }
      if (context.tags) {
        Object.entries(context.tags).forEach(([key, value]) => {
          scope.setTag(key, value)
        })
      }
      if (context.extra) {
        Object.entries(context.extra).forEach(([key, value]) => {
          scope.setExtra(key, value)
        })
      }
      if (context.level) {
        scope.setLevel(context.level)
      }
      Sentry.captureException(error)
    })
  } else {
    Sentry.captureException(error)
  }
}

/**
 * Capture a message (for non-error events)
 */
export function captureMessage(
  message: string,
  level: Sentry.SeverityLevel = 'info',
  context?: {
    tags?: Record<string, string>
    extra?: Record<string, any>
  }
) {
  if (context) {
    Sentry.withScope((scope) => {
      if (context.tags) {
        Object.entries(context.tags).forEach(([key, value]) => {
          scope.setTag(key, value)
        })
      }
      if (context.extra) {
        Object.entries(context.extra).forEach(([key, value]) => {
          scope.setExtra(key, value)
        })
      }
      scope.setLevel(level)
      Sentry.captureMessage(message)
    })
  } else {
    Sentry.captureMessage(message, level)
  }
}

/**
 * Set user context for error tracking
 */
export function setUser(user: {
  id: string
  email?: string
  username?: string
  [key: string]: any
}) {
  Sentry.setUser(user)
}

/**
 * Clear user context (e.g., on logout)
 */
export function clearUser() {
  Sentry.setUser(null)
}

/**
 * Add breadcrumb for debugging
 */
export function addBreadcrumb(
  message: string,
  category: string,
  data?: Record<string, any>
) {
  Sentry.addBreadcrumb({
    message,
    category,
    data,
    level: 'info',
    timestamp: Date.now() / 1000,
  })
}

/**
 * Start a transaction for performance monitoring
 */
export function startTransaction(name: string, op: string) {
  return Sentry.startTransaction({
    name,
    op,
  })
}

/**
 * Wrap an async function with error tracking
 */
export function withErrorTracking<T extends (...args: any[]) => Promise<any>>(
  fn: T,
  context?: {
    tags?: Record<string, string>
    name?: string
  }
): T {
  return (async (...args: Parameters<T>) => {
    try {
      return await fn(...args)
    } catch (error) {
      captureError(error as Error, {
        tags: context?.tags,
        extra: {
          functionName: context?.name || fn.name,
          arguments: args,
        },
      })
      throw error
    }
  }) as T
}

File: api/src/lib/monitoring/sentry.py
"""
Sentry Error Tracking Configuration - Backend

Centralized error tracking and monitoring for FastAPI backend using Sentry.io
Captures errors, performance metrics, and request context.

Setup:
1. Sign up at https://sentry.io (Free tier: 5K events/month)
2. Create project and get DSN
3. Add SENTRY_DSN to .env.local
4. Call init_sentry() at application startup
5. Install: pip install sentry-sdk[fastapi]
"""

import os
from typing import Any, Dict, Optional

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration
from sentry_sdk.integrations.sqlalchemy import SqlAlchemyIntegration


def init_sentry() -> None:
    """
    Initialize Sentry for error tracking.

    Call this at application startup (e.g., in main.py).
    """
    dsn = os.getenv("SENTRY_DSN")

    if not dsn:
        print("[Sentry] DSN not configured, error tracking disabled")
        return

    environment = os.getenv("SENTRY_ENVIRONMENT", os.getenv("NODE_ENV", "development"))

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        # Sample rates
        traces_sample_rate=0.1 if environment == "production" else 1.0,
        profiles_sample_rate=0.1 if environment == "production" else 1.0,

        # Integrations
        integrations=[
            FastApiIntegration(
                transaction_style="endpoint",  # Group by endpoint not URL
            ),
            StarletteIntegration(
                transaction_style="endpoint",
            ),
            SqlAlchemyIntegration(),
        ],

        # Filter sensitive data
        before_send=before_send_filter,

        # Ignore breadcrumbs callback
        before_breadcrumb=before_breadcrumb_filter,
    )


def before_send_filter(event: Dict[str, Any], hint: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Filter sensitive data before sending to Sentry.

    Args:
        event: Sentry event dict
        hint: Additional context

    Returns:
        Filtered event or None to drop
    """
    # Remove sensitive request data
    if "request" in event:
        request = event["request"]

        # Remove authorization headers
        if "headers" in request:
            headers = request["headers"]
            headers.pop("authorization", None)
            headers.pop("cookie", None)
            headers.pop("x-api-key", None)

        # Sanitize query string
        if "query_string" in request:
            query = request["query_string"]
            if query:
                # Remove token parameters
                sanitized = query.replace("token=", "token=[REDACTED]")
                sanitized = sanitized.replace("api_key=", "api_key=[REDACTED]")
                request["query_string"] = sanitized

    # Remove sensitive extra data
    if "extra" in event:
        extra = event["extra"]
        # Remove password fields
        for key in list(extra.keys()):
            if "password" in key.lower() or "secret" in key.lower():
                extra[key] = "[REDACTED]"

    # Remove sensitive context
    if "contexts" in event:
        contexts = event["contexts"]

        # Sanitize database context
        if "database" in contexts:
            db = contexts["database"]
            if "connection_string" in db:
                db["connection_string"] = "[REDACTED]"

        # Sanitize Stripe/payment context
        if "stripe" in contexts:
            stripe = contexts["stripe"]
            stripe.pop("secret_key", None)
            stripe.pop("api_key", None)

    return event


def before_breadcrumb_filter(crumb: Dict[str, Any], hint: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Filter breadcrumbs before adding to Sentry.

    Args:
        crumb: Breadcrumb dict
        hint: Additional context

    Returns:
        Filtered breadcrumb or None to drop
    """
    # Don't track database query breadcrumbs (too noisy)
    if crumb.get("category") == "query":
        return None

    # Don't track HTTP requests to analytics/tracking services
    if crumb.get("category") == "httplib":
        url = crumb.get("data", {}).get("url", "")
        if any(x in url for x in ["analytics", "tracking", "segment", "mixpanel"]):
            return None

    return crumb


def capture_error(
    error: Exception,
    context: Optional[Dict[str, Any]] = None,
    user: Optional[Dict[str, str]] = None,
    tags: Optional[Dict[str, str]] = None,
    extra: Optional[Dict[str, Any]] = None,
    level: str = "error"
) -> None:
    """
    Capture an error with additional context.

    Args:
        error: The exception to capture
        context: Additional context dict
        user: User information (id, email, username)
        tags: Tags for filtering/grouping
        extra: Extra data for debugging
        level: Severity level (debug, info, warning, error, fatal)
    """
    with sentry_sdk.push_scope() as scope:
        if user:
            scope.set_user(user)

        if tags:
            for key, value in tags.items():
                scope.set_tag(key, value)

        if extra:
            for key, value in extra.items():
                scope.set_extra(key, value)

        if context:
            scope.set_context("custom", context)

        scope.set_level(level)
        sentry_sdk.capture_exception(error)


def capture_message(
    message: str,
    level: str = "info",
    tags: Optional[Dict[str, str]] = None,
    extra: Optional[Dict[str, Any]] = None,
) -> None:
    """
    Capture a message (for non-error events).

    Args:
        message: Message to capture
        level: Severity level
        tags: Tags for filtering/grouping
        extra: Extra data
    """
    with sentry_sdk.push_scope() as scope:
        if tags:
            for key, value in tags.items():
                scope.set_tag(key, value)

        if extra:
            for key, value in extra.items():
                scope.set_extra(key, value)

        scope.set_level(level)
        sentry_sdk.capture_message(message)


def set_user(user_id: str, email: Optional[str] = None, username: Optional[str] = None) -> None:
    """
    Set user context for error tracking.

    Args:
        user_id: User ID
        email: User email (optional)
        username: Username (optional)
    """
    sentry_sdk.set_user({
        "id": user_id,
        "email": email,
        "username": username,
    })


def clear_user() -> None:
    """Clear user context (e.g., on logout)."""
    sentry_sdk.set_user(None)


def add_breadcrumb(message: str, category: str, data: Optional[Dict[str, Any]] = None, level: str = "info") -> None:
    """
    Add a breadcrumb for debugging.

    Args:
        message: Breadcrumb message
        category: Category for grouping
        data: Additional data
        level: Severity level
    """
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        data=data or {},
        level=level,
    )


def start_transaction(name: str, op: str) -> Any:
    """
    Start a transaction for performance monitoring.

    Args:
        name: Transaction name
        op: Operation type (e.g., "http.server", "db.query")

    Returns:
        Transaction object
    """
    return sentry_sdk.start_transaction(name=name, op=op)
